### DISCUSSION TITLE
Hydration error with inline tags

### DISCUSSION THREAD

**Discussion by khash** [markdoc/markdoc#185] [Bug report] [answered]
*Using NextJS*

I'm trying to build a "hint" tag, similar to this from https://stripe.com/docs/connect/collect-then-transfer-guide

<img width="318" alt="image" src="https://user-images.githubusercontent.com/1121822/186906960-01936975-f8d1-488d-966d-5a8f84978516.png">

This component, should be usable within any text (like in `code`) and accept anything as the content of the popover (links, images, etc).

I think this is a simple use case and below is how I've reproduced it. I've removed all code that does that for simplicity in the question:

```js
// hint.markdoc.js
export default {
  attributes: {
    caption: { type: String },
  },
  render: ({ caption, children }) => <div><span>{caption}</span><span>{children}</span></div>,
};
```

```
// test.md
This is a {% hint caption="hint"%}generated by the tag{% /hint %} as an example
```

This throws a NextJS Hydration error https://nextjs.org/docs/messages/react-hydration-error. 

However the same code with the following works fine:

```markdoc
This is a 
{% hint caption="hint"%}
generated by the tag
{% /hint %}
as an example
```

Which is not desirable as the text is broken into three lines.

To understand what's going on, I logged the `children` prop in the render method, the inline, shows the text and the non-inline version shows the `p` element (see below). I'm unsure as to if that's relevant, but I thought it might be worth mentioning. 

```
{
  '$$typeof': Symbol(react.element),
  type: 'p',
  key: null,
  ref: null,
  props: {
    children: 'generated by the tag'
  },
  _owner: null,
  _store: {}
}
```

When ignoring the errors the result is rendered as below, because each part of the text is wrapped in a `p`:

```
This is a {% hint caption="hint"%}generated by the tag{% /hint %} as an example
```

<img width="285" alt="image" src="https://user-images.githubusercontent.com/1121822/186900817-2efaadbc-03b9-430d-bc05-e454cb50568f.png">

**Comment by mfix-stripe** [comment, marked as answer]
@khash I'll look into the server mismatch error more thoroughly (Cc @rpaul-stripe if you have any ideas). 

As for how your component is rendering, the reason it is on multiple lines is because you are rendering a `div`, which is a block level element. If you replace this with a `span`, everything will render correctly. When you put the `hint` tag on a line by itself, Markdoc will treat it as a block level element, and it will be wrapped with a `p` tag, causing the content to render how you want it in that case:

<img width="201" alt="image" src="https://user-images.githubusercontent.com/62121649/186952712-22b09f43-04bb-4630-b1d7-ecb50bef8855.png">

**Reply by khash** [reply]
Thank you @mfix-stripe changing the `div` to `span` fixed the multi-line issue.

**Reply by khash** [reply]
@mfix-stripe @rpaul-stripe FYI, the issue with the Hydration error is still there. Not making the tag inline works (it returns a `p` tag) but using inline returns just a string and raises the hydration error. I wonder how this works on your own docs at stripe

**Reply by mfix-stripe** [reply]
@khash totally. Would you mind creating a new _issue_ for the hydration error? That would be very helpful. Thanks @khash!

**Reply by mfix-stripe** [reply]
Just opened https://github.com/markdoc/markdoc/issues/188 👍

**Reply by khash** [reply]
Thank you @mfix-stripe please let me know if I can help
